src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -482,32 +482,17 @@ void AtlasEngine::SetWarningCallback(std::function<void(HRESULT, wil::zwstring_v
     // commit 9e86c98 (PR #16196), because it showed that it's definitely not due to FindFamilyName() failing.
     //
     // The workaround is to catch the exception and retry it with our nearby fonts manually loaded in.
+    HRESULT hr = _updateFont(fontInfoDesired, fontInfo, features, axes);
+
     if constexpr (Feature_NearbyFontLoading::IsEnabled())
     {
-        try
-        {
-            _updateFont(fontInfoDesired, fontInfo, features, axes);
-            return S_OK;
-        }
-        CATCH_LOG();
-
-        // _resolveFontMetrics() checks `_api.s->font->fontCollection` for a pre-existing font collection,
-        // before falling back to using the system font collection. This way we can inject our custom one.
-        // Doing it this way is a bit hacky, but it does have the benefit that we can cache a font collection
-        // instance across font changes, like when zooming the font size rapidly using the scroll wheel.
-        try
+        if (FAILED(hr) && _updateWithNearbyFontCollection())
         {
-            _api.s.write()->font.write()->fontCollection = FontCache::GetCached();
+            hr = _updateFont(fontInfoDesired, fontInfo, features, axes);
         }
-        CATCH_LOG();
     }
 
-    try
-    {
-        _updateFont(fontInfoDesired, fontInfo, features, axes);
-        return S_OK;
-    }
-    CATCH_RETURN();
+    return hr;

```

Comment:
HRESULT makes it simpler to write the `if failed and necessary to retry` condition IMO.